### PR TITLE
shared files for the whole function-app

### PR DESCRIPTION
**Pull request by yonatandt**
Signed-off-by: YonatanTal <yonatald@gmail.com>

### CODE REVIEW

File: shared/event/event.model.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import * as mongoose from 'mongoose';
+
+import { IEvent, Status } from './event.interface';
```
Comment (ShragaUser) [inline shared/event/event.model.ts:3]: Am I missing something? Is Status even defined in event.interface?

File: shared/event/event.model.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+const EventSchema = new mongoose.Schema({
```
Comment (ShragaUser) [inline shared/event/event.model.ts:5]: Also - why even use mongoose at all? Why not the actual mongodb driver? Using interfaces for validation in application level?

File: shared/malshab/malshab.interface.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    gender?: Gender;
+    birthDate?: Date;
+    medicalProfile?: MedicalProfile;
+    kaba?: number;
+    dapar?: number;
+    continuousDapar?: number;
+    zadak?: number;
+    kahas?: number;
+    major?: number;
+    aptitudeTest?: AptitudeTest;
```
Comment (ShragaUser) [inline shared/malshab/malshab.interface.ts:56]: What is this? If it's a number shouldn't it have a "translated" property or be an enum?

File: shared/malshab/malshab.interface.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    languages?: Language[];
+    englishGrade?: number;
+    tzalagAgrade?: number;
+    tzalagBgrade?: number;
+    isPassedKogTirgum?: boolean;
+    isPersonalSurvey?: boolean;
+    isReligiousSchool?: boolean;
+    shakimSchoolAverage?: number;
+    isArabist?: boolean;
+    pdfQuestionnaire?: string;
```
Comment (ShragaUser) [inline shared/malshab/malshab.interface.ts:80]: This is a url? If so - better to explicitly say so in property name

File: shared/mongoose.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import * as mongoose from 'mongoose';
```
Comment (ShragaUser) [inline shared/mongoose.ts:1]: Disconnection handling?
Error handling? 
Logs?

File: shared/event/event.model.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    location: String,
+    interviewersIds: [mongoose.Types.ObjectId],
+    status: {
+        type: String,
+        enum: Status,
+    },
+    url: String,
+});
+
+const EventModel = mongoose.model<IEvent & mongoose.Document>(process.env.EVENT_COLLECTION_NAME, EventSchema);
```
Comment (ShragaUser) [inline shared/event/event.model.ts:18]: Shouldn't there be a default value for configurable elements? Maybe you should use a config file that holds all global configurations and handles them correctly?

**Review by ShragaUser — CHANGES_REQUESTED** [review]

File: shared/event/event.interface.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+export interface Event {
```
Comment (carmeltsadok) [inline shared/event/event.interface.ts:1]:
```suggestion
export interface IEvent {
```

File: shared/error/error.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+export default class ApplicationError extends Error {
+    constructor(message: string) {
```
Comment (carmeltsadok) [inline shared/error/error.ts:2]: What about error code?

File: shared/malshab/malshab.interface.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
+export type Gender = 1 | 2;
+export type MedicalProfile = 21 | 24 | 25 | 45 | 64 | 72 | 82 | 97;
+
+interface AptitudeTest {
```
Comment (carmeltsadok) [inline shared/malshab/malshab.interface.ts:4]:
```suggestion
interface IAptitudeTest {
```

File: shared/malshab/malshab.interface.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    madadKeshevMitmasheh: number;
+    madadKeshevSelektivi: number;
+    madadPikud: number;
+    madadAvodatTsevet: number;
+    madadHashkaVehatmada: number;
+    madadBashlutVebagrut: number;
+    madadHitnahagutMisgartit: number;
+}
+
+export interface Address {
```
Comment (carmeltsadok) [inline shared/malshab/malshab.interface.ts:21]:
```suggestion
export interface IAddress {
```

File: shared/malshab/malshab.interface.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    street: string;
+    houseNumber: number;
+    cityId: number;
+    cityName: string;
+}
+
+interface Language {
```
Comment (carmeltsadok) [inline shared/malshab/malshab.interface.ts:28]:
```suggestion
interface ILanguage {
```

File: shared/malshab/malshab.interface.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    languageId: number;
+    languageName: string;
+    languageLevel: number;
+    languageLevelDesc: string;
+}
+
+interface Grade {
```
Comment (carmeltsadok) [inline shared/malshab/malshab.interface.ts:35]:
```suggestion
interface IGrade {
```

File: shared/malshab/malshab.interface.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    examId: number;
+    examName: string;
+    grade: number;
+}
+
+export default interface Malshab {
```
Comment (carmeltsadok) [inline shared/malshab/malshab.interface.ts:41]:
```suggestion
export default interface IMalshab {
```

File: shared/unit/unit.interface.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+export default interface Unit {
```
Comment (carmeltsadok) [inline shared/unit/unit.interface.ts:1]:
```suggestion
export default interface IUnit {
```

Comment (carmeltsadok) [inline shared/unit/unit.interface.ts:1]: Unit model is missing

File: shared/nodesGroup/nodeGroups.interface.ts [hunk 15]
```diff
@@ -0,0 +1,6 @@
+export default interface NodesGroup {
```
Comment (carmeltsadok) [inline shared/nodesGroup/nodeGroups.interface.ts:1]:
```suggestion
export default interface INodesGroup {
```

Comment (carmeltsadok) [inline shared/nodesGroup/nodeGroups.interface.ts:1]: Change file name to nodesGroup.interface.ts.
NodesGroup model is missing.

**Review by carmeltsadok — CHANGES_REQUESTED** [review]

File: shared/error/error.ts (hunk 8 above)
Comment (yonatandt) [reply shared/error/error.ts:2]: I think it is not mandatory and is not really needed.

Comment (ShragaUser) [reply shared/error/error.ts:2]: I think supporting added error code is a good idea. Might not be mandatory, but it is relevant...

File: shared/malshab/malshab.interface.ts [hunk 16]
```diff
@@ -0,0 +1,83 @@
+export type Gender = 1 | 2;
```
Comment (yubbi99) [inline shared/malshab/malshab.interface.ts:1]: dont limit the malshab gender

**Review by lustigy96 — CHANGES_REQUESTED** [review]

File: shared/malshab/malshab.interface.ts [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+    homePhoneNumber?: string;
+    identityNumber: string;
+    imigrationDate?: Date;
+    isArabist?: boolean;
+    isBereaved?: boolean;
+    isOnlyChild?: boolean;
+    isPassedKogTirgum?: boolean;
+    isPersonalSurvey?: boolean;
+    isReligiousSchool?: boolean;
+    kaba?: number;
```
[the remark this answers is not in the document]
Comment (yonatandt) [reply shared/malshab/malshab.interface.ts:63]: There is no need for it. We are not supposed to validate that. We trust our input completely.

File: shared/malshab/malshab.interface.ts [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+    languageName: string;
+}
+
+export default interface Malshab {
+    addresses?: Address[];
+    aptitudeTest?: AptitudeTest;
+    birthCountry?: string;
+    birthDate?: Date;
+    continuousDapar?: number;
+    dapar?: number;
```
[the remark this answers is not in the document]
Comment (yonatandt) [reply shared/malshab/malshab.interface.ts:46]: There is no need for it. We are not supposed to validate that. We trust our input completely.

File: shared/malshab/malshab.interface.ts [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+interface Language {
+    languageId: number;
+    languageLevel: number;
+    languageLevelDesc: string;
+    languageName: string;
+}
+
+export default interface Malshab {
```
[the remark this answers is not in the document]
Comment (yonatandt) [reply shared/malshab/malshab.interface.ts:40]: That is not true. The default export doesn't have to be the only exported value, it's just the default one.

File: shared/malshab/malshab.interface.ts [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+    addresses?: Address[];
+    aptitudeTest?: AptitudeTest;
+    birthCountry?: string;
+    birthDate?: Date;
+    continuousDapar?: number;
+    dapar?: number;
+    email?: string;
+    englishGrade?: number;
+    firstName?: string;
+    gender?: number;
```
[the remark this answers is not in the document]
Comment (yonatandt) [reply shared/malshab/malshab.interface.ts:50]: Good luck putting all the genders in an enum 😉

File: shared/mongoose.ts [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+let dbInstance: mongoose.Connection;
+
+const connectToMongo = async () => {
+    try {
+        await mongoose.connect(config.mongo.uri, {
+            useNewUrlParser: true,
+            useUnifiedTopology: true,
+            useFindAndModify: false,
+            useCreateIndex: true,
+        });
```
[the remark this answers is not in the document]
Comment (yonatandt) [reply shared/mongoose.ts:14]: Why should we?
It can work without it.

File: shared/event/event.interface.ts (hunk 7 above)
Comment (AdarPascal) [reply shared/event/event.interface.ts:6]: Not a global interface naming convention standard. Only Yesodot's habit.

**Review by AdarPascal — APPROVED** [review]

File: shared/event/event.interface.ts (hunk 7 above)
Comment (yonatandt) [reply shared/event/event.interface.ts:6]: That's not only a Yesodot's standard. In C# for example, interfaces usually have an `I` prefix.
But it seems that most developers tend not to use it, as in [this](https://stackoverflow.com/questions/2814805/java-interfaces-implementation-naming-convention) [this](https://stackoverflow.com/questions/541912/interface-naming-in-java) and [this](https://softwareengineering.stackexchange.com/questions/117348/should-interface-names-begin-with-an-i-prefix).

File: shared/malshab/malshab.interface.ts [hunk 22]
```diff
@@ -0,0 +1,81 @@
+export type MedicalProfile = 21 | 24 | 25 | 45 | 64 | 72 | 82 | 97;
+
```
Comment (lustigy96) [inline shared/malshab/malshab.interface.ts:2]: I think there should be an enum of Gender (and not type) because it is clearer.

**Review by lustigy96 — CHANGES_REQUESTED** [review]

**Review by carmeltsadok — APPROVED** [review]

File: shared/error/error.ts [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+ */
+export class FunctionError extends Error {
+    /**
+     * @param message - A message describing the error.
+     * @param functionName - The name of the azure function in which the error occurred.
+     * @param info - Additional info on the error.
+     */
+    constructor(message: string, functionName: string, info: string) {
+        super(message);
+        if (!message) {
```
Comment (carmeltsadok) [inline shared/error/error.ts:20]: The message is required, so it will never go there.

File: shared/unit/unit.interface.ts [hunk 24]
```diff
@@ -0,0 +1,4 @@
...
+    id: string;
```
Comment (carmeltsadok) [inline shared/unit/unit.interface.ts:2]: For type validation, id shouldn't exist
```suggestion
```

File: shared/unit/unit.model.ts [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+import config from '../../config';
+
+const { collectionNames } = config.mongo;
+
+const unitSchema = new Schema({
+    id: {
+        type: String,
+        required: true,
+        index: true,
+    },
```
Comment (carmeltsadok) [inline shared/unit/unit.model.ts:13]: For type validation, id shouldn't exist
```suggestion
```

**Review by carmeltsadok — CHANGES_REQUESTED** [review]

File: shared/unit/unit.model.ts [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+    name: {
+        type: String,
+        required: true,
+        index: true,
+    },
```
Comment (carmeltsadok) [inline shared/unit/unit.model.ts:18]:
```suggestion
    name: {
        type: String,
        unique: true,
        required: true,
        index: true,
    },
```

**Review by carmeltsadok — CHANGES_REQUESTED** [review]

File: shared/mongoose.ts [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+        dbInstance = mongoose.connection;
+    } catch {
+        console.log('failed to connect to MongoDB...');
+    }
+};
+
+dbInstance.on('connecting', () => {
```
Comment (carmeltsadok) [inline shared/mongoose.ts:21]: There is a bug when connecting. dbInstance is undefined

**Review by carmeltsadok — CHANGES_REQUESTED** [review]

**Review by maykremer — CHANGES_REQUESTED** [review]

File: shared/authWrapper.ts [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+ */
+const handleUserToken = (token: string): User | null => {
+    jwt.verify(token, config.auth.secret, (err, decoded) => {
+        if (err) {
+            // later, log error
+            return null;
+        }
+        return formatUserFromToken(decoded);
+    });
+    return null;
```
Comment (carmeltsadok) [inline shared/authWrapper.ts:32]:
```suggestion
    let formattedUser = null;
    jwt.verify(token, config.auth.secret, (err, decoded) => {
        if (err) {
            // later, log error
        }
        formattedUser = formatUserFromToken(decoded);
    });
    return formattedUser;
```

File: shared/authWrapper.ts [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+        if (header && header.split(' ')[1]) {
+            const user = handleUserToken(header.split(' ')[1]);
+            if (user) {
+                await func(context, req, user);
+            }
+        }
+        context.res = {
+            status: 401,
+        };
+    };
```
Comment (carmeltsadok) [inline shared/authWrapper.ts:53]:
```suggestion
        if (header && header.split(' ')[1]) {
            const user = handleUserToken(header.split(' ')[1]);
            if (user) {
                await func(context, req, user);
            } else {
                context.res = {
                    status: 401,
                };
            }
        } else {
            context.res = {
                status: 401,
            };
        }
```

File: shared/nodesGroup/nodeGroups.interface.ts [hunk 30]
```diff
@@ -0,0 +1,6 @@
...
+    id: string;
```
Comment (maykremer) [inline shared/nodesGroup/nodeGroups.interface.ts]: nodes group interface should be:
    id: string;
    name: string;
    unitId: string;
    usersIds: string[];
    nodesIds: string[];

File: shared/nodesGroup/nodesGroup.interface.ts [hunk 31]
```diff
@@ -0,0 +1,6 @@
+export default interface NodesGroup {
+    id: string;
+    name: string;
+    unitId: string;
+    usersIds: string[];
+}
```
Comment (maykremer) [inline shared/nodesGroup/nodesGroup.interface.ts:6]: Nodes group interface should include : nodesIds: string[];

File: shared/nodesGroup/nodesGroup.model.ts [hunk 32]
```diff
@@ -0,0 +1,32 @@
...
+        type: String,
+        required: true,
+        index: true,
+    },
+    usersIds: [
+        {
+            type: Schema.Types.ObjectId,
+            ref: collectionNames.user,
+        },
+    ],
```
Comment (maykremer) [inline shared/nodesGroup/nodesGroup.model.ts:29]: Nodes group model should also include :
  nodesIds: [
        {
            type: String,
            ref: collectionNames.node,
        },
    ],

File: shared/node/node.model.ts [hunk 33]
```diff
@@ -0,0 +1,38 @@
...
+        required: true,
+        unique: true,
+        index: true,
+    },
+    type: {
+        type: String,
+        enum: NodeType,
+        default: NodeType.Unknown,
+    },
+    nodeGroupId: {
```
Comment (lustigy96) [inline shared/node/node.model.ts:26]:
```suggestion
    nodesGroupId: {
```

spelling according to the interface

File: shared/node/node.model.ts [hunk 34]
```diff
@@ -0,0 +1,38 @@
...
+        type: Schema.Types.ObjectId,
+        ref: collectionNames.nodesGroup,
+    },
+    unitId: {
+        type: Schema.Types.ObjectId,
+        ref: collectionNames.unit,
+    },
+});
+
+// TODO: <unitId,name> should be unique.
```
Comment (lustigy96) [inline shared/node/node.model.ts:36]: why unitId should  be unique? the might be multiple nodes per unit

**Review by lustigy96 — CHANGES_REQUESTED** [review]

File: shared/nodesGroup/nodesGroup.interface.ts [hunk 35]
```diff
@@ -0,0 +1,6 @@
+export default interface NodesGroup {
+    id: string;
+    name: string;
+    unitId: string;
+    usersIds: string[];
```
Comment (maykremer) [inline shared/nodesGroup/nodesGroup.interface.ts:5]: usersIds type should be array of objectId.

Comment (yonatandt) [reply shared/nodesGroup/nodesGroup.interface.ts:5]: I think that string should be enough for our dedicated use.
Why do you believe that `string` is not enough?

File: shared/node/node.interface.ts [hunk 36]
```diff
@@ -0,0 +1,14 @@
+export enum NodeType {
+    Interview = 'INTERVIEW',
+    Exam = 'EXAM',
+    Physical = 'PHYSICAL',
+    Unknown = 'UNKNOWN',
+}
+
+export interface Node {
```
[the remark this answers is not in the document]
Comment (yonatandt) [reply shared/node/node.interface.ts:8]: We have decided to not use the `I` prefix in an interface's name.

File: shared/user/user.model.ts [hunk 37]
```diff
@@ -0,0 +1,30 @@
...
+        index: true,
+    },
+    role: {
+        type: String,
+        enum: Role,
+        required: true,
+    },
+    unitId: {
+        type: Schema.Types.ObjectId,
+        ref: collectionNames.unit,
```
[the remark this answers is not in the document]
Comment (yonatandt) [reply shared/user/user.model.ts:26]: A user cannot be in more than one unit. But there are users that don't even have a unit, like a technical user.

File: shared/malshab/malshab.interface.ts (hunk 22 above)
Comment (yonatandt) [reply shared/malshab/malshab.interface.ts:2]: I truly believe there is no reason to. Either way, we get that as an input without validation.

File: shared/error/error.ts (hunk 23 above)
Comment (yonatandt) [reply shared/error/error.ts:20]: message can be falsy (undefined or an empty string)

File: shared/unit/unit.interface.ts (hunk 24 above)
Comment (yonatandt) [reply shared/unit/unit.interface.ts]: Why should it be that way?

File: shared/event/event.interface.ts [hunk 38]
```diff
@@ -0,0 +1,21 @@
...
+        id: string;
+        firstName: string;
+        lastName: string;
+    };
+    time: Date;
+    location: string;
+    interviewersIds: string[];
+    url?: string;
+    results?: EventResult[];
+}
```
[the remark this answers is not in the document]
Comment (yonatandt) [reply shared/event/event.interface.ts:21]: It is not in the excalidraw.

File: shared/event/event.model.ts [hunk 39]
```diff
@@ -0,0 +1,58 @@
...
+import { Event } from './event.interface';
+
+import config from '../../config';
+
+const { collectionNames } = config.mongo;
+
+const eventSchema = new Schema({
+    node: {
+        id: {
+            type: Schema.Types.ObjectId,
```
[the remark this answers is not in the document]
Comment (yonatandt) [reply shared/event/event.model.ts:12]: Its a refernce to the node, therefore it should be an objectId.

File: shared/event/event.model.ts [hunk 40]
```diff
@@ -0,0 +1,58 @@
...
+            ref: collectionNames.node,
+        },
+        name: {
+            type: String,
+            required: true,
+        },
+    },
+    malshabShort: {
+        id: {
+            type: Schema.Types.ObjectId,
```
[the remark this answers is not in the document]
Comment (yonatandt) [reply shared/event/event.model.ts:22]: Its a refernce to the malshab, therefore it should be an objectId.

File: shared/node/node.model.ts (hunk 34 above)
Comment (yonatandt) [reply shared/node/node.model.ts:36]: The couple `<unitId, name>` together should be unique

File: shared/nodesGroup/nodesGroup.interface.ts (hunk 31 above)
Comment (yonatandt) [reply shared/nodesGroup/nodesGroup.interface.ts:6]: The node already includes an `nodesGroupId`

File: shared/nodesGroup/nodesGroup.model.ts (hunk 32 above)
Comment (yonatandt) [reply shared/nodesGroup/nodesGroup.model.ts:29]: The node already includes an `nodesGroupId`

File: shared/unit/unit.interface.ts [hunk 41]
```diff
@@ -0,0 +1,6 @@
+import { Types} from 'mongoose';
+
+export default interface Unit {
+    _id?: Types.ObjectId,
+    name: string;
+}
```
Comment (carmeltsadok) [inline shared/unit/unit.interface.ts:6]:
```suggestion
export default interface Unit {
    _id?: string,
    name: string;
}
```

**Review by carmeltsadok — CHANGES_REQUESTED** [review]

File: shared/authWrapper.ts [hunk 42]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * wrapWithAuth - Wraps an HTTP Trigger Azure function with authentication handling.
+ * @param func - An HTTP Trigger Azure function that also gets a User object.
+ * @returns a function that wraps func with authentication handling that either send
+ * the authenticated user to func or returns a Unauthenticated error to the client.
+ */
+export default (func: AzureFunction): AzureFunction => {
+    return async (context: Context, req: HttpRequest) => {
+        const header = req.headers.Authorization;
```
Comment (maykremer) [inline shared/authWrapper.ts:44]:
```suggestion
        const header = req.headers.authorization;
```

**Review by maykremer — CHANGES_REQUESTED** [review]

**Review by carmeltsadok — DISMISSED** [review]

**Review by maykremer — APPROVED** [review]

**Review by lustigy96 — APPROVED** [review]

**Pull request merged**